workload: run directly on storage engines

ajkr opened the pull request:
This PR forks the `workload run` subcommand to create a `workload kv-run`
that runs the specified workload on RocksDB. The goal is to be able to compare
following three scenarios with a single measurement tool:
    
- RocksDB performance when used via libroach API so it includes CGo overhead. Can be done using `workload kv-run`.
- Pebble performance. Will be done by a future tool in Pebble repo that invokes the function `KvOpsRun` that is introduced in this PR.
- Cockroach performance when SQL queries are derived from basic KV operations. Can be done using the existing `workload kv`.

Release note: None

cockroach-teamcity commented:
<!-- Reviewable:start -->
This change is [<img src="https://reviewable.io/review_button.png" height="34" align="absmiddle" alt="Reviewable"/>](https://reviewable.io/reviews/cockroachdb/cockroach/35721)
<!-- Reviewable:end -->

danhhz commented:
This is awesome! Happy to see it

> expands the scope of workload beyond its original purpose of hitting Cockroach with queries

Not at all! I've actually envisioned from the beginning that workload would handle things like setting up an `engine.Engine` with some initial dataset and having an associated kv-level queryload is a natural extension of that. Can I suggest the following structure which cleans up some of the dep issues (and will give us some more flexibility for storage engine bechmarking in the future)?

In `pkg/workload`:

Introduce a new interface parallel to `Opser` that any workloads with an associated engine queryload will implement. Initially this will only be `kv`. We can probably think of a better name for this, but for the sake of this explanation, call it `KVOpser`.

In `pkg/workload/cli`:

Introduce an interface that abstracts over rocksdb and pebble. Maybe `engine.ReadWriter`, maybe something simpler at first. Then fork the `cockroach workload run` cli subcommand to a new `cockroach workload kv-run` (again, probably a better name for this). Have this new subcommand set up either rocksdb or pebble (as requested by flags) and do whatever bechmarking is appropriate given some `KVOpser` implementor. Most of this code will be identical to the `run` stuff at first, but I can see it diverging over time as engine benchmarking may want to be in control of its own destiny.

Possible modification to this:

In the short term, I imagine that we're interested in keeping pebble in the petermattis repo for easier iteration. If that's the case, then the command I mention in `pkg/workload/cli` could be made a thin wrapper around some logic that lives in `pkg/workload`. Then we could vendor just the workload bits of the cockroach repo into the pebble repo and make a second thin wrapper cli command there. Workload's deps are pretty light and we could even trim them a bit more if needed. That way, one could easily run the same benchmarks on rocksdb and pebble, but without requiring that they both be in the same repo. We'd have to updated the vendored copy of workload in pebble if the workload impl changed, but hopefully this is a rare.

danhhz commented:
> TODO: the KV operations it synthesizes when run in raw KV mode should be
as similar as possible to the KV operations Cockroach executes when this
tool runs in SQL mode. Currently they're a bit different.

IMPORT already has paved the way a bit for turning SQL things into KV things. `rowConverter` uses the regular INSERT code to turn rows of datums into the kv representation. Turning arbitrary INSERT/UPDATE queries into kv ops is hard, but a whitelist of simple ones could be tractable. Definitely a big enough project that what you're doing here (manually doing the translation) is the right thing for the short term.

https://github.com/cockroachdb/cockroach/blob/b8303ba074c8a6c7cd8698b55da2cb47f0934cde/pkg/ccl/importccl/read_import_proc.go#L199-L349

tbg commented:
I'm happy to see this, too. Just want to avoid workload getting lots of dep
bloat, but it seems that Dan has you covered.

ajkr commented:
Thanks a lot for all the detailed suggestions! I will try them out and let
you know how it goes.

On Thu, Mar 14, 2019, 8:47 AM Tobias Grieger <notifications@github.com>
wrote:

> I'm happy to see this, too. Just want to avoid workload getting lots of dep
> bloat, but it seems that Dan has you covered.
>
> —
> You are receiving this because you authored the thread.
> Reply to this email directly, view it on GitHub
> <https://github.com/cockroachdb/cockroach/pull/35721#issuecomment-472928360>,
> or mute the thread
> <https://github.com/notifications/unsubscribe-auth/AEjxSuCg8NCgiAnDqW45D88yBPMjm1tpks5vWm8CgaJpZM4bzCN8>
> .
>

ajkr committed:
workload/kv: run directly on storage engines

This PR splits the `workload kv` tool into a SQL mode and a raw KV
mode that runs on either Pebble or RocksDB. Then we can compare the
following three scenarios with a single measurement tool:

- RocksDB performance when used via libroach API so it includes CGo overhead
- Pebble performance
- Cockroach performance when SQL queries are derived from basic KV operations

Release note: None

ajkr commented:
@danhhz updated per your comments. Let me know if you have any more suggestions. I'm especially curious what you think about the interface between "workload/cli/kv_run.go" and "workload/kv_ops_runner.go". It feels like the division between them is pretty arbitrary right now, although I didn't have much ideas to do things differently.

ajkr commented:
Here are a few more open problems related to this PR that I've been thinking about but don't have answers to:

- Was `engine.Engine` a reasonable choice for requiring `KvOpsRun()` callers to provide? I chose that instead of `engine.ReadWriter` due to the need for the `NewBatch()` API.
- Since I reused the logic in "workload/kv/kv.go" to implement both `run kv` and `kv-run kv`, the CLI arguments for both commands are the same. That means the `kv-run kv` inherits a lot of irrelevant arguments, like `--secondary-index`, `--opt`, `--splits`, etc. They should be no-ops if set but it's a little bit confusing.
- A command using `kv-run` looks something like `./bin/workload kv-run kv --store-dir ./tmp-rocks --duration 5s --read-percent 80`. The `kv-run kv` part looks a bit confusing.

ajkr commented:
I'm having some difficulty getting just the workload bits vendored. Maybe it has too many dependencies or maybe I'm doing it wrong.

Commands:
```
$ GO111MODULE=on go get github.com/cockroachdb/cockroach/pkg/workload
$ GO111MODULE=on go mod vendor
```

Result: "go.mod", "go.sum", and "vendor/modules.txt" all appear to be 10x larger.

ajkr commented:
> I'm having some difficulty getting just the workload bits vendored. Maybe it has too many dependencies or maybe I'm doing it wrong.

Tried the same thing using a package with no dependencies (`pkg/util/stringencoding`). It did not pull in any extra dependencies, so I guess the focus now can be on reducing `pkg/workload`'s dependencies.

petermattis commented:
Are you trying to vendor workload into Pebble? That seems fraught. Perhaps I’m misunderstanding the vendoring you’re trying to do.

ajkr commented:
Yes. It worked besides pulling in way too many things to the vendor directory and go.mod. I didn't really have a sense of what is better - making Pebble depend on workload or making workload depend on Pebble - so tried both. The benefit of the former is we could have a benchmark tool readily available in Pebble repo, and the drawback is the dependency mess. With the latter, we'd have to use a tool in cockroach repo for measuring both, but the benefit is dependencies would be understandable. Actually I still don't know which is better. Good news is we have the code for both. Bad news is I have a feeling we might end up using neither and just copying the code, unless @danhhz intervenes and enlighten us :).

petermattis commented:
If we vendor workload into Pebble it will be awkward to vendor Pebble into Cockroach. If we tie Pebble into workload within the Cockroach repo we’re externalizing the benchmarking of Pebble in a way that feels unnatural to me. I think copying what we want from workload into Pebble is the right approach, though I’m open to hearing other arguments or approaches.

ajkr commented:
I thought about it a bit and don't really think this will be a circular dependency if we do it right. Conceptually it'd look like this:

- Pebble bench command depends on Pebble engine and `cockroach/pkg/workload`
- `cockroach/pkg/workload/cli` depends on CockroachDB, RocksDB, and `cockroach/pkg/workload`
- `cockroach/pkg/workload` depends on standard library and utilities only (not currently the case but theoretically possible I think?)

The benefit is we avoid copying code and having to maintain very similar code in two repos. I started copying and it's hard to decide what to keep and what to omit (e.g., `workload kv` supports zipfian distribution - do we copy that too?).

ajkr commented:
BTW, I plan to complete the trilogy and submit the copied code as a Pebble PR either way. Just wanted to write down some thoughts on picking from the three not-so-great options.

petermattis commented:
> I thought about it a bit and don't really think this will be a circular dependency if we do it right.

I think we could get this to work between the packages involved, but the circular dependency between the repos will remain awkward.

> BTW, I plan to complete the trilogy and submit the copied code as a Pebble PR either way. Just wanted to write down some thoughts on picking from the three not-so-great options.

👍 My 2 cents: I don't mind copying code from workload for this purpose. You haven't mentioned DRY, but I think that advice is overrated.

ajkr commented:
> I think we could get this to work between the packages involved, but the circular dependency between the repos will remain awkward.

If the dependencies are not conceptually circular, but expressing them requires a circular dependency between repos, could it indicate the repo setup isn't ideal (at least for this particular situation)? If the Pebble/Cockroach repos were merged, or if common Cockroach utilities were in a third repo, it'd avoid this dissonance. Do you have a prediction of how much of the Cockroach repo's utilities we'll need or an opinion on whether exploring these ideas further is worthwhile?

petermattis commented:
> If the dependencies are not conceptually circular, but expressing them requires a circular dependency between repos, could it indicate the repo setup isn't ideal (at least for this particular situation)? If the Pebble/Cockroach repos were merged, or if common Cockroach utilities were in a third repo, it'd avoid this dissonance. 

Merging the Pebble/Cockroach repos would make Pebble much less easy to use separate from Cockroach. Perhaps that is a good thing. 

Pulling utilities out of the cockroach repo to a location they can be shared will then make development of those utilities a bit more burdensome. In some cases that is ok (such as `testutils/datadriven` mentioned below). I'm pragmatic, but intuitive about this decision. If we identify some cockroach utilities that would be super useful in Pebble, by all means let's share them. But I don't want to share for the motivation of DRY. Sometimes copying is ok too.

> Do you have a prediction of how much of the Cockroach repo's utilities we'll need or an opinion on whether exploring these ideas further is worthwhile?

The only thing we've grabbed so far is `testutils/datadriven` (aka `internal/datadriven`). That was recently pulled out of the cockroach repo and it is probably worthwhile to use a shared version. There are no other utility libraries that are on my radar to use. For example, we wouldn't want to use `util/log`, because the logging for a component like Pebble should be pluggable externally.

danhhz commented:
IIRC, vendoring workload into pebble was originally my suggestion. Some of the unfortunate deps of pkg/workload should be easy to fix, but there's an annoying one which includes some un-checked-in generated code, which means that you currently cannot simply go get/mod it.

I'm also generally okay with copying code. My worry with that in this case is around whether we want the output of the rocksdb benchmark to be comparable with the pebble one (which I'm assuming is the whole point). In that case, we'd need to be careful to copy over any changes that happen that would affect these numbers, which seems brittle.

Not sure if this was what was being suggested, but just in case: I don't think moving workload to a third repo is a good idea yet. It's intended to be general purpose enough that we could do that eventually and I think it's a cool idea, but it's currently changing enough that making every workload PR into 2 PRs in 2 repos seems like too much overhead.

My opinion is that, in the short term, we should work through the unfortunate workload deps or simply copy in the code (I don't have a strong opinion which). If/when we vendor pebble into cockroach, we could resolve the repo cycle by moving all the benchmarking into the cockroach repo. It would mean the standalone pebble repo wouldn't have its own benchmarking, but perhaps that's an acceptable tradeoff?

I will readily admit my monorepo bias, but I'm intrigued by biting the bullet and moving pebble into the cockroach repo's pkg tree. This certainly adds process to pebble PRs but makes this benchmarking work trivial. It also may come with other benefits from our existing cockroach build tooling (auto issues for test failures, inclusion in the daily github digest, backport, etc).

If we then ever wanted to make pebble back into its own project, I was part of an effort at Foursquare to do exactly this. Foursquare had a private repo but had internal tools that we wanted to open source. We ended up building a nightly job that strips a single folder out of the monorepo and makes it into a new one, keeping relevant commit/blame history. It works surprisingly well: https://github.com/foursquare/fsqio. The foursquare opensource repo is also a monorepo, but there's no reason our version couldn't make a cockroach->pebble repo and a cockroach->workload repo, or whatever else we think would be more generally useful to the go community.
